osbs/build/user_params_meta.py

```diff
@@ -0,0 +1,176 @@
+"""
+Copyright (c) 2020 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+
+from __future__ import absolute_import, unicode_literals
+
+import six
+
+from osbs.exceptions import OsbsValidationException
+
+
+class BuildParam(object):
+    """
+    One parameter of a spec.
+
+    Works like a data descriptor (similar to `property`), should be defined as
+    a class attribute on classes that inherit from BuildParamsBase.
+
+    When defined as a class attribute, the name of the attribute should match
+    the name given to the BuildParam in __init__(), e.g.:
+
+    >>> class BuildParams(BuildParamsBase)
+    >>>     x = BuildParam("x")
+    >>>     y = BuildParam("y")
+
+    BuildParamsBase and classes that inherit from it will validate this during
+    class creation.
+    """
+
+    def __init__(self, name, default=None, required=False, include_in_json=True):
+        """
+        Define a BuildParam.
+
+        Do NOT use mutable values as default, same issues apply as when using
+        mutable defaults for method params.
+
+        :param name: str, name of param, should match name of class attribute
+        :param default: any immutable object, default value of param
+        :param required: bool, is the param allowed to be None?
+        :param include_in_json: bool, include the param in to_json() output?
+        """
+        self._name = name
+        self._default = default
+        self._required = required
+        self._include_in_json = include_in_json
+        self._mangled_name = "_{self.__class__.__name__}__{self.name}".format(self=self)
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def required(self):
+        return self._required
+
+    @property
+    def include_in_json(self):
+        return self._include_in_json
+
+    def __repr__(self):
+        return "{self.__class__.__name__}({self.name!r})".format(self=self)
+
+    def __get__(self, obj, objtype=None):
+        try:
+            # Bypass potential __getattr__ redefinition in class
+            return object.__getattribute__(obj, self._mangled_name)
```

Comment:
Could you elaborate on why simply using something like `self.val` is not sufficient? I don't fully understand why we're setting this attribute on the owner object, e.g. `BuildParamsBase`.

Reply:
One of the main goals of this refactoring was to change `BuildParam` to a class attribute instead of an instance attribute (basically the opposite of [this](https://github.com/containerbuildsystem/osbs-client/commit/a2b06d4e0b1d4b7c0241cf2500409d55008476e7), as you mentioned).

If the value was stored in the `BuildParam` instance, it would be shared among all instances of `BuildParamsBase`. Example:

```python
class SharedValue(object):
    def __init__(self):
        self._value = None

    def __get__(self, obj, objtype=None):
        return self._value

    def __set__(self, obj, value):
        self._value = value


class Foo(object):
    x = SharedValue()


foo = Foo()
bar = Foo()

foo.x = 1
print(foo.x, bar.x)  # 1 1
bar.x = 2
print(foo.x, bar.x)  # 2 2
```

That is exactly what we *do not* want, so the attribute has to be set on the owner to prevent this behaviour. See this test: https://github.com/chmeliik/osbs-client/blob/0b4e0fe81cc21372055bdfae647becd5a63f6912/tests/build_/test_user_params_meta.py#L206-L219

Reply:
So we want a class attribute to behave like an instance attribute? It's unclear to me why we want this. Python's syntax for accessing such attributes is exactly the same. The only difference I can think of is where those attributes are defined:

```python
class Foobar():
  spam = BuildParam()

  def __init__(self, ...):
    sef.spam = ...

# versus

class Foobar():
  def __init__(self, ...):
    self.spam = BuildParam(...)
```

Those values need to be initialized anyways in `__init__`, so why not also define them there?

Reply:
I've misunderstood how data descriptors are working under the hood. The example I posted above is invalid and does not work as expected. They must be set as a class attributes.